[configuration] Allow to specify a list of CPUs to bind H2O to

deweerdt opened the pull request:
Modify `num-threads` to accept a sequence. If passed a sequence, `num-threads` will treat it as a list of CPUs to bind to, instantiation one thread per CPU.

kazuho commented:
Thank you for the PR. I think this is a nice enhancement.

That said, I think we might want to consider extending `num_threads`. Because I'd assume that what you'd actually want to do is to pin each of our worker threads to each of the specified CPU core. Or do you have a different use-case in mind?

Regarding listing the CPU cores, I think we might want to use a YAML sequence here. Just naming all the CPU core IDs in numbers could be fine, or if we want to support range expressions, we could support YAML expression like `[0, 3-6, 9]` (in which we parse "3-6" as a single range).

deweerdt commented:
That's a good idea, thanks for suggesting it. IIUC, `num-threads` would take two kinds of values:
- a single integer, in which case it designates a the number of threads to spawn
- a sequence, in which case the number of threads will be specified by the number of CPUs in the sequence, and each thread will be assigned to one CPU

Is that correct?

deweerdt commented:
@kazuho how does 2147a98 look?

kazuho commented on src/main.c:
```diff
@@ -1220,15 +1222,51 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
+    *num_threads += 1;
+    (*cpu_list) = realloc(*cpu_list, sizeof(**cpu_list) * (*num_threads + 1));
+    (*cpu_list)[*num_threads - 1] = cpu;
+    (*cpu_list)[*num_threads] = -1;
```
Are you concerned about having overruns here?

I ask this because, *if* you are concerned about that, *then* I might suggest doing a refactor. Replace `conf.num_threads` and `conf.cpu_list` with a `H2O_VECTOR(int)`. The size of the vector indicates the number of threads, and each value of the vector indicates the CPU id (if non-negative) to be pinned to, or if it's not pinned (if -1). By taking that approach, there would be no fear of losing synchronization between `num_threads` and `cpu_list`, and you'd also be able to use things like `h2o_vector_reserve`.

I am not saying that we should make such change, but it's generally the case that I prefer avoiding having extra complexity for making the code "fail-safe."

deweerdt commented:
Thanks for your review @kazuho ! I believe 71a3b71 addresses all your comments so far.

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
 static int on_config_num_threads(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node)
 {
-    if (h2o_configurator_scanf(cmd, node, "%zu", &conf.num_threads) != 0)
+    size_t num_threads = 0;
```
Could you please move the variable declaration inside `if (node->type == YOML_TYPE_SCALAR) {`? I think that the only block that uses the variable.

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
-    if (h2o_configurator_scanf(cmd, node, "%zu", &conf.num_threads) != 0)
+    size_t num_threads = 0;
+    free(conf.thread_map.entries);
+    memset(&conf.thread_map, 0, sizeof(conf.thread_map));
```
While these two lines work, simply setting `conf.thread_map.size` to zero would be sufficient.

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
+    free(conf.thread_map.entries);
+    memset(&conf.thread_map, 0, sizeof(conf.thread_map));
+    if (node->type == YOML_TYPE_SCALAR) {
+        int i;
```
Could it be changed to `size_t` considering that the variables used in conjunction with `i` are `size_t`?

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
+            conf.thread_map.entries[conf.thread_map.size++] = -1;
+    } else if (node->type == YOML_TYPE_SEQUENCE) {
+#ifndef H2O_HAS_PTHREAD_SETAFFINITY_NP
+        fprintf(stderr, "[error] Can't handle a CPU list, this platform doesn't support `pthread_setaffinity_np`\n");
```
I'd appreciate it if you could use `h2o_configurator_errprintf`. Unless we do so, our users might have hard time figuring out how to resolve the issue.

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
-    if (conf.num_threads == 0) {
+#endif
+        /* a sequence is treated as a list of CPUs to bind to, one per thread to instantiate */
+        int i;
```
`size_t`

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
+        for (i = 0; i < node->data.sequence.size; i++) {
+            char *cpu_spec;
+            if (node->data.sequence.elements[i]->type != YOML_TYPE_SCALAR) {
+                h2o_configurator_errprintf(cmd, node, "CPUs in cpu sequence must be a scalar");
```
We should better designate `node->data.sequence.elements[i]` as the offending node. The same goes for the other invocations of `h2o_configurator_errprintf` found below.

kazuho commented on src/main.c:
```diff
@@ -1222,9 +1222,56 @@ static int on_config_max_connections(h2o_configurator_command_t *cmd, h2o_config
+                conf.thread_map.entries[conf.thread_map.size++] = cpu_num;
+            } else {
+                int pos, cpu_low, cpu_high, cpu_num;
+                if (sscanf(cpu_spec, "%d-%d%n", &cpu_low, &cpu_high, &pos) != 2 || pos != strlen(cpu_spec) || cpu_low < 0 || cpu_high < 0) {
```
Would it make sense to use `%u` instead? We would not want to see something like `-3-1` being accepted.

kazuho commented on src/main.c:
```diff
@@ -1987,9 +2046,12 @@ static void setup_configurators(void)
     const char *cmd = argv[0], *opt_config_file = H2O_TO_STR(H2O_CONFIG_PATH);
-    int error_log_fd = -1;
+    int n, error_log_fd = -1;
+    size_t num_threads = h2o_numproc();
```
Nitpick, but we might want to name the variable `num_procs`, because it indicates the number of CPU cores, not the number of threads in any particular way.

deweerdt commented:
Thanks for the review, @kazuho I believe ab3d087 addresses all the comments you've made so far. I've also used an intermediate variable instead of accessing `node->data.sequence.elements[i]` directly.

kazuho commented:
Thank you for the changes. I've merged the PR with some tweaks. I believe that the changes are not disputable (mostly refactor and editorial changes), but please let me know if you have concerns.

deweerdt commented:
> Thank you for the changes. I've merged the PR with some tweaks. I believe that the changes are not disputable (mostly refactor and editorial changes), but please let me know if you have concerns.

Changes look great to me, thank you.
